activerecord/lib/arel/visitors/postgresql.rb

```diff
@@ -78,6 +78,26 @@ def visit_Arel_Nodes_IsDistinctFrom(o, collector)
           visit o.right, collector
         end
 
+        # Postgres-specific implementation that uses `col = any('{1,2}')` instead of `col IN (1,2)`
+        # to avoid pg_stat_statements churn
+        def visit_Arel_Nodes_HomogeneousIn(o, collector)
+          oid = ActiveRecord::ConnectionAdapters::PostgreSQL::OID
+          case o.attribute.type_caster
+          when ActiveRecord::Encryption::EncryptedAttributeType, oid::Bytea, oid::Jsonb
+            return super
```

Comment:
These data types don't currently work. Maybe because they get encoded twice? (see #49050)

The main risk of this PR is that there are other data types that are broken, but don't have test coverage right now.

A less risky option would be to only use the new logic for types known to work (number, string, boolean).

Reply:
Rails test coverage is usually pretty good. But if you’re concerned enough to mention it, then explicitly opting in the types is probably a good idea.

Reply:
I'm not concerned, just trying to make review easier. The test suite didn't catch the fact that JSONB doesn't work.

It would be good to review all of the Postgres data types someone would expect this sort of query to work for, to see if any others are broken.

Reply:
Wouldn't be possible to only opt-in types known to be working properly (and tested) for now and fallback to `super` for rest and iterate over this list if needed later? Using `ANY` for list of numbers or "strings" should cover majority of the problems already if I understand it well.

Reply:
While it's possible to opt-in to only the types known to work, that would mean opting into almost every type available. I don't think it's enough to only support strings and numbers, because there are so many unique Postgres data types that deserve the optimization this PR adds (e.g. inet, uuid, money, range).

I wonder if we should focus on getting #49050 merged -- that may remove the need to have a branching path here.

CC @JohnAnon9771, the author of that PR

Reply:
I've added #49050 to this branch locally, with some additions to this custom `HomogeneousIn` implementation:
```rb
type_caster = oid::Array.new(o.attribute.type_caster, ",")
casted_values = o.values.map do |raw_value|
  o.attribute.type_caster.serialize(raw_value) if o.attribute.type_caster.serializable?(raw_value)
end
casted_values.compact!
pg_encoder = PG::TextEncoder::Array.new name: "#{type_caster.type}[]", delimiter: ","
values = [oid::Array::Data.new(pg_encoder, casted_values)]
proc_for_binds = -> value { Arel::Nodes::HomogeneousIn::BindAttribute.new(o.attribute.name, value) }
```
That fixes queries for encrypted columns, but unfortunately doesn't fix binary or JSONB columns.

I think I'll need to write a set of tests that covers all Postgres data types that Rails supports, so we can know for sure which types need to fall back to `IN`.

Reply:
> I've added #49050 to this branch locally, with some additions to this custom `HomogeneousIn` implementation:
> 
> ```ruby
> type_caster = oid::Array.new(o.attribute.type_caster, ",")
> casted_values = o.values.map do |raw_value|
>   o.attribute.type_caster.serialize(raw_value) if o.attribute.type_caster.serializable?(raw_value)
> end
> casted_values.compact!
> pg_encoder = PG::TextEncoder::Array.new name: "#{type_caster.type}[]", delimiter: ","
> values = [oid::Array::Data.new(pg_encoder, casted_values)]
> proc_for_binds = -> value { Arel::Nodes::HomogeneousIn::BindAttribute.new(o.attribute.name, value) }
> ```
> 
> That fixes queries for encrypted columns, but unfortunately doesn't fix binary or JSONB columns.
> 
> I think I'll need to write a set of tests that covers all Postgres data types that Rails supports, so we can know for sure which types need to fall back to `IN`.

Thank you for testing my PR to look at this, could you tell me how the searches for jsonb and binary turned out?

Reply:
I haven't looked into the issue with binary, but it looks like JSONB does technically work, but I was confused because the test only checks the bind params; it doesn't actually perform a query to see if the expected rows are found. @JohnAnon9771 you may want to update that test to be more comprehensive.

The code is generating string-encoded JSON inside the array, which I would expect Postgres to treat as a JSON string instead of a JSON object.
```sql
json_options" = any('{"{\"a\":1}","{\"b\":2}"}')
```

However, it turns out that Postgres accepts two different formats for JSONB arrays:

```sql
-- string-encoded
select a->>'a' from unnest('{"{\"a\":2}","{\"a\":3}"}'::jsonb[]) _(a);
 ?column? 
----------
 2
 3

-- JSON literal, with { and } escaped
select a->>'a' from unnest('{\{\"a\":2\},\{\"a\":3\}}'::jsonb[]) _(a);
 ?column? 
----------
 2
 3
```

While a non-array JSONB value is correctly treated as a string:
```sql
select '"{\"a\":2}"'::jsonb ->> 'a';
 ?column? 
----------
 
```

I'm not sure what's happening here. @msakrejda any ideas?

Reply:
@seanlinsley I think this is due to the Postgres array input syntax: https://www.postgresql.org/docs/current/arrays.html#ARRAYS-INPUT -- an array value starts with curly brackets, and then individual elements can either be quoted (as in the first example, requiring you to escape any nested double quotes inside the array element values) or not quoted (requiring you to escape any curly brackets and commas inside the array element values).

So you're actually dealing with arrays of JSON objects in both cases, due to how the array input syntax works.